packages/openapi3/test/test-return-types.ts

```diff
@@ -0,0 +1,238 @@
+    ok(res.paths["/"].put.responses["201"].content["application/json"].schema);
+  });
+
+  it("defines responses with headers and status codes", async () => {
+    const res = await openApiFor(
+      `
+      model ETagHeader {
+        @header eTag: string;
+      }
+      model CreatedResponse {
+        @statusCode code: "201";
+      }
+      model Key {
+        key: string;
+      }
+      @route("/")
+      namespace root {
+        @put
+        op create(): { ...CreatedResponse, ...ETagHeader, @body body: Key};
+      }
+      `
+    );
+    ok(res.paths["/"].put.responses["201"]);
+    ok(res.paths["/"].put.responses["201"].headers["e-tag"]);
+    deepStrictEqual(res.paths["/"].put.responses["201"].content["application/json"].schema, {
+      $ref: "#/components/schemas/Key",
+    });
+  });
+
+  it("defines separate responses for each variant of a union return type", async () => {
+    const res = await openApiFor(
+      `
+      @doc("Error")
+      model Error {
+        @statusCode _: "default";
+        code: int32;
+        message: string;
+      }
+      model Key {
+        key: string;
+      }
+      @route("/")
+      namespace root {
+        @get
+        op read(): Key | Error;
+      }
+      `
+    );
+    ok(res.paths["/"].get.responses["200"]);
+    ok(res.components.schemas.Key);
+    deepStrictEqual(res.paths["/"].get.responses["200"].content["application/json"].schema, {
+      $ref: "#/components/schemas/Key",
+    });
+    ok(res.components.schemas.Error);
+    deepStrictEqual(res.paths["/"].get.responses["default"].content["application/json"].schema, {
+      $ref: "#/components/schemas/Error",
+    });
+  });
+
+  it("defines the response media type from the content-type header if present", async () => {
+    const res = await openApiFor(
+      `
+      @doc("Error")
+      model Error {
+        @statusCode _: "default";
+        code: int32;
+        message: string;
+      }
+      model TextPlain {
+        @header contentType: "text/plain";
+      }
+      model Key {
+        key: string;
+      }
+      @route("/")
+      namespace root {
+        @get
+        // Note: & takes precedence over |
+        op read(): Key & TextPlain | Error;
+      }
+      `
+    );
+    ok(res.paths["/"].get.responses["200"]);
+    ok(res.paths["/"].get.responses["200"].content["text/plain"].schema);
+    ok(res.components.schemas.Error);
+    deepStrictEqual(res.paths["/"].get.responses["default"].content["application/json"].schema, {
+      $ref: "#/components/schemas/Error",
+    });
+  });
+
+  it("returns diagnostics for duplicate body decorator", async () => {
+    const res = await checkFor(
+      `
+      model Foo {
+        foo: string;
+      }
+      model Bar {
+        bar: string;
+      }
+      @route("/")
+      namespace root {
+        @get
+        op read(): { @body body1: Foo, @body body2: Bar };
+      }
+      `
+    );
+    strictEqual(getDiagnostic("duplicate-body", [...res]).length, 1);
```

Comment:
Feels like this helper doesn't make test looks very understandable. I'd say we either
1. stick with just testing the diagnostics array directly. I have been doing things like:
```ts
strictEqual(diagnostics.length, 1); 
strictEqual(diagnostics[0].code, "@cadl-lang/rest/operation-param-duplicate-type");
strictEqual(diagnostics[0].message, "Param multiParam has multiple types: [query, path]");
```
This lets us test the message as well which could be different and require interpolation

2. Have maybe a new helper method `assertDiagnostic(code: string, message?: string)`

Reply:
Yeah, I like having assertions for a particular diagnostic over the diagnostic array.  It may be somewhat fragile to rely on the order that diagnostics are reported.  It might be useful to have an `assertDiagnostics(diagnostics: Diagnostic[], expected: {code: string, message?: string}[])` to verify that *only* these diagnostics appear.

Reply:
I think it is fragile only if you are testing multiple diagnostics at once which you shouldn't IMO unless it is the goal.
If you have multiple diagnostics then teh assertDiagnostics could also just be as confusing of an error depending on how it tries to find the diagnostics.

Reply:
Okay.  I borrowed this code from somewhere but I'm happy to change it.